pkg/plugins/kafka/source.go

```diff
@@ -21,67 +21,65 @@ import (
 	"time"
 
 	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/plugin/sdk"
 	"github.com/conduitio/conduit/pkg/plugins"
-	"github.com/conduitio/conduit/pkg/record"
 	"github.com/segmentio/kafka-go"
 )
 
 type Source struct {
+	sdk.UnimplementedSource
+
 	Consumer         Consumer
 	Config           Config
-	lastPositionRead record.Position
+	lastPositionRead sdk.Position
+}
+
+func NewSource() sdk.Source {
+	return &Source{}
 }
 
-func (s *Source) Open(ctx context.Context, cfg plugins.Config) error {
-	fmt.Println("Opening a Kafka Source...")
-	parsed, err := Parse(cfg.Settings)
+func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
+	fmt.Println("Configuring a Kafka Source...")
+	parsed, err := Parse(cfg)
 	if err != nil {
 		return cerrors.Errorf("config is invalid: %w", err)
 	}
 	s.Config = parsed
+	return nil
+}
 
+func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
 	client, err := NewConsumer()
 	if err != nil {
 		return cerrors.Errorf("failed to create Kafka client: %w", err)
 	}
-
 	s.Consumer = client
-	return nil
-}
 
-func (s *Source) Teardown() error {
-	fmt.Println("Tearing down a Kafka Source...")
-	s.Consumer.Close()
-	return nil
-}
-
-func (s *Source) Validate(cfg plugins.Config) error {
-	_, err := Parse(cfg.Settings)
-	return err
-}
-
-func (s *Source) Read(ctx context.Context, position record.Position) (record.Record, error) {
-	err := s.startFrom(position)
+	err = s.startFrom(pos)
 	if err != nil {
-		return record.Record{}, cerrors.Errorf("couldn't start from position: %w", err)
+		return cerrors.Errorf("couldn't start from position: %w", err)
 	}
 
+	return nil
+}
+
+func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
 	message, kafkaPos, err := s.Consumer.Get(ctx)
 	if err != nil {
-		return record.Record{}, cerrors.Errorf("failed getting a message %w", err)
+		return sdk.Record{}, cerrors.Errorf("failed getting a message %w", err)
 	}
 	if message == nil {
-		return record.Record{}, plugins.ErrEndData
+		return sdk.Record{}, plugins.ErrEndData
```

Comment:
This error should now be `sdk.ErrBackoffRetry`. In the future we should rewrite `s.Consumer.Get` to block until either the context is cancelled or a message is read.

Reply:
👍 Done.